### PR TITLE
SQL: Use a QueryProcessManager subclass

### PR DESCRIPTION
**Pull request by mia-pi-git**
Draft because there's a critical bug in this code that causes the subprocesses to crash when started. No idea why. No errors are logged.

### CODE REVIEW

**Comment by mia-pi-git** [comment]
I have since determined that this _only_ happens to the modlog database. Unsure of why.

**Comment by mia-pi-git** [comment]
I fixed it. The bug was that in the child process, Modlog was not being instantiated, so the receiving child process wasn't there.

File: lib/sql.ts [hunk 1]
```diff
@@ -16,310 +14,252 @@ export interface SQLOptions {
...
-	}
-	all(statement: Statement, data: DataType = {}) {
-		const num = typeof statement === 'number' ? statement : this.statements.get(statement);
-		if (num === undefined || ![...this.statements.values()].includes(num)) {
-			throw new Error(`Prepare a statement before using another database function with SQLDatabase.prepare.`);
+	setupDatabase() {
+		if (this.dbReady) return;
+		this.dbReady = true;
+		const {file, extension} = this.options;
+		// console.log(`loading database ${file}${extension ? ` and extension ${extension}` : ''}`);
```
Comment (AnnikaCodes) [inline lib/sql.ts:172]:
```suggestion
```

File: lib/sql.ts [hunk 2]
```diff
@@ -16,310 +14,252 @@ export interface SQLOptions {
...
-			process.send!(`STATEMENTS\n${JSON.stringify(Object.fromEntries(statements))}\n`);
-		}
-		if (onDatabaseStart) {
-			onDatabaseStart(database);
+		if (result === SQLiteErrors.DatabaseNotFound) {
+			throw new Error("SQLite database was queried, but Config.usesqlite is off");
 		}
+		return result;
 	}
-	database?.pragma(`foreign_keys=on`);
```
Comment (AnnikaCodes) [inline lib/sql.ts:235]: does this line go anywhere? was it actually necessary?

File: server/modlog/index.ts [hunk 3]
```diff
@@ -78,52 +77,62 @@ export interface ModlogEntry {
...
+	altsInsertionQuery: SQL.Statement | null = null;
+	renameQuery: SQL.Statement | null = null;
+	globalPunishmentsSearchQuery: SQL.Statement | null = null;
 
-	constructor(databasePath: string, sqliteOptions?: Partial<SQLOptions>) {
+	constructor(databasePath: string, options: Partial<SQL.Options>) {
 		this.queuedEntries = [];
 		this.databaseReady = false;
+		const dbExists = FS(databasePath).existsSync();
+		// if (process.send) console.log('trying db?');
```
Comment (AnnikaCodes) [inline server/modlog/index.ts:102]:
```suggestion
```

File: server/modlog/index.ts [hunk 4]
```diff
@@ -364,7 +373,7 @@ export class Modlog {
 		query += ` ${sortAndLimit.query}`;
 		args.push(...sortAndLimit.args);
 
-		return {statement: await this.database.prepare(query), queryText: query, args};
+		return {statement: (await this.database.prepare(query))!.toString(), queryText: query, args};
```
Comment (AnnikaCodes) [inline server/modlog/index.ts:376]: Uhhhhhhh, why do we prepare it and then convert it to a string?

File: lib/sql.ts (hunk 2 above)
Comment (mia-pi-git) [reply lib/sql.ts:235]: Don't think it was needed. Consumers can choose themselves.

Comment (AnnikaCodes) [reply lib/sql.ts:235]: OK.

**Comment by AnnikaCodes** [comment]
Alright, let's hope this doesn't break anything.

**Pull request merged**